content/advent-2017/messaging-framework.md

```diff
@@ -0,0 +1,709 @@
++++
+author = ["Orr Chen"]
+title = "Simple messaging framework using Go TCP server and Kafka"
+linktitle = "messaging framework"
+date = 2017-12-31T00:00:00Z
+series = ["Advent 2017"]
++++  
+  
+![System diagram](/static/postimages/advent-2017/messaging-framework/system-diagram.png)  
+I needed to create a simple framework to provide my endpoint devices ( doesn't matter which platform they run on ) the option to send and receive messages from my backend.  
+I require those messages to be managed by a message broker so that they can be processed in an asynchronous way.  
+The system contains 4 main layers, this article section is mainly about the first one:  
+1. **TCP servers** - Needs to keep maintain as many as possible concurrent tcp sockets with the endpoints. all of the endpoints messages will be processed on a different layer through the message broker. This keeps the TCP servers layer very thin and effective and to keep as many concurrent connection as possible, and Go is a good pick for it ( see [this article](https://medium.freecodecamp.org/million-websockets-and-go-cc58418460bb))  
+2. **Message broker** - Responsible for delivering the messages between the TCP servers layer and the workers layer. I chose [Apache Kafka](https://kafka.apache.org/) for that purpose.  
+3. **Workers layer** - will process the messages through services exposed in the backend layer.  
+4. **Backed services layer** - An encapsulation of services requires by your application such as DB, Authentication, Logging, external APIs and more.  
+  
+So, this Go Server:  
+1. communicates with its endpoint clients by TCP sockets.  
+2. queues the messages in the message broker.  
+3. receives back messages from the broker after they were processed and send response acknowledgment and/or errors to the TCP clients.  
+
+The full source code is available in : https://github.com/orrchen/go-messaging  
+I have also included a Dockerfile and a build script to push the image to your Docker repository.  
+Special thanks to the great go Kafka [sarama library from Shopify](https://github.com/Shopify/sarama).
+
+_The article is divided to sections representing the components of the system. Each component should be decoupled from the others in a way that if you are interested in reading about just one of components it should be straight forward._
```

Comment:
s/way that if you are interested in reading about just one of components it should be straight forward/way that allows you to read about a single component in a straight forward manner